walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,155 @@
+   * @return the ArrayList of TimeRanges representing the time blocks where events occur
+   */
+  private ArrayList<TimeRange> mergeOverlappingEvents(ArrayList<Event> eventList) {
+    ArrayList<TimeRange> mergedEventTimes = new ArrayList<>();
+    
+    if (eventList.isEmpty()) {
+      return mergedEventTimes;
+    }
+
+    mergedEventTimes.add(eventList.get(0).getWhen());
+
+    for (int i = 1; i < eventList.size(); i++) {
+      // For two adjacent events, check if they overlap
+      TimeRange lastTimeRange = mergedEventTimes.get(mergedEventTimes.size() - 1);
+      TimeRange currEventTime = eventList.get(i).getWhen(); 
+      if (lastTimeRange.overlaps(currEventTime)) {
+        // If they do, create new TimeRange with combined time and add to mergedEventTimes,
+        // replacing the original TimeRange the new event merged with
+        int newEnd = Math.max(lastTimeRange.end(), currEventTime.end());
+        TimeRange combinedTime = TimeRange.fromStartEnd(lastTimeRange.start(), newEnd, false);
+        mergedEventTimes.remove(mergedEventTimes.size() - 1);
+        mergedEventTimes.add(combinedTime);
+      }
+      else {
+        // Otherwise, add the current event to the list
+        mergedEventTimes.add(currEventTime);
+      }
+    }
+    return mergedEventTimes;
+  }
+
+  /**
+   * Given a sorted list of TimeRanges where events take place, return a list of the availability
+   * (no event) times from the start of the day (00:00) to the end of the day (23:59) that are 
+   * at least of the given duration.
+   * @param eventTimes the sorted list of TimeRanges that represent the time periods in the day
+   *     when events take place
+   * @param duration the minimum length of the period of availability 
+   */
+  private ArrayList<TimeRange> findAvailability(ArrayList<TimeRange> eventTimes, long duration) {
+    
+    ArrayList<TimeRange> availability = new ArrayList<TimeRange>();
+    
+    // If there are no meetings, then the whole day is free
+    if (eventTimes.isEmpty()) {
+      availability.add(TimeRange.WHOLE_DAY);
+      return availability;
+    }
+
+    // Start at the beginning of the day
+    int timeslotStart = TimeRange.START_OF_DAY;
+
+    for (int i = 0; i < eventTimes.size(); i++) {
+      // If the gap between timeSlotstart and the start of the next event is >= meeting duration,
+      // add to availability
+      TimeRange currEvent = eventTimes.get(i);
+      if (currEvent.start() - timeslotStart >= duration) {
+        TimeRange timeslot = TimeRange.fromStartEnd(timeslotStart, currEvent.start(), false);
+        availability.add(timeslot);
+      }
+      // the next potential meeting time will start from the end of the current event
+      timeslotStart = currEvent.end();
+    }
+
+    // Finally, add the timeslot from the end of the last event to the end of the day
+    TimeRange lastEvent = eventTimes.get(eventTimes.size() - 1);
+    if (TimeRange.END_OF_DAY - lastEvent.end() >= duration) {
```

Comment:
Instead of starting from `lastEvent.end()`, you could just start from `timeslotStart` - this would handle the no meetings case properly, so you can remove that edge case at the beginning of this function. Then, this function would also properly handle `duration` greater than an entire day, so you wouldn't need the special case on line 40 too.

Reply:
Ooh yes, didn't pick that up. Thank you!